### PR TITLE
Restructure docs

### PR DESCRIPTION
**Pull request by spawnia**

### CODE REVIEW

**Review by kitikhunp — CHANGES_REQUESTED** [review]
[0](https://1)

**Comment by kitikhunp** [comment]
Https://1

**Comment by kitikhunp** [comment]
0

File: docs/best-practices.md [hunk 1]
```diff
@@ -1,9 +1,52 @@
...
+namespace MyApp;
+
+class TypeRegistry
+{
+    private $myAType;
+    private $myBType;
+    
+    public function myAType()
+    {
+        return $this->myAType ?: ($this->myAType = new MyAType($this));
```
Comment (simPod) [inline docs/best-practices.md:19]: we should use `??`

File: docs/best-practices.md [hunk 2]
```diff
@@ -1,9 +1,52 @@
...
+    {
+        parent::__construct([
+            'fields' => [
+                'b' => $types->myBType()                
+            ]
+        ]);
+    }
+}
+```
+Obviously, you can automate this registry as you wish to reduce boilerplate or even
```
Comment (simPod) [inline docs/best-practices.md:48]: maybe add line break before

**Review by simPod — APPROVED** [review]

File: docs/best-practices.md (hunk 2 above)
Comment (spawnia) [reply docs/best-practices.md:48]: That will be handled by `prettier` once we settled its setup in https://github.com/webonyx/graphql-php/pull/808.

File: docs/schema-definition.md [hunk 3]
```diff
@@ -116,60 +114,64 @@ In this case, it is recommended to pass **typeLoader** option to schema construc
...
 
-$registry = new Types();
+$typeRegistry = new TypeRegistry();
 
 $schema = new Schema([
-    'query' => $registry->get('Query'),
-    'typeLoader' => function($name) use ($registry) {
-        return $registry->get($name);
-    }
+    'query' => $typeRegistry('Query'),
```
Comment (simPod) [inline docs/schema-definition.md:159]: TBH I liked `get` much better. This reads like `type registry query WHAT?`. IMO too fancy which is not necessary.

File: docs/schema-definition.md [hunk 4]
```diff
@@ -116,60 +114,64 @@ In this case, it is recommended to pass **typeLoader** option to schema construc
...
 
-    public function get($name)
+    public function __invoke(string $name): Type
     {
-        if (!isset($this->types[$name])) {
-            $this->types[$name] = $this->{$name}();
-        }
-        return $this->types[$name];
+        return $this->types[$name]
+            ??= $this->{$name}();
```
Comment (simPod) [inline docs/schema-definition.md:135]: maybe put it on single line?

File: docs/schema-definition.md (hunk 3 above)
Comment (spawnia) [reply docs/schema-definition.md:159]: The line below becomes much nicer though and shaves off a level of indirection.

Comment (spawnia) [reply docs/schema-definition.md:159]: It is objectively worse in terms of performance though.

Comment (simPod) [reply docs/schema-definition.md:159]: I guess it adds micro overhead. But for docs it should rather be verbose.

Comment (spawnia) [reply docs/schema-definition.md]: Fine.

**Pull request merged**